x-pack/plugins/fleet/server/services/security/uninstall_token_service/index.ts

```diff
@@ -0,0 +1,412 @@
+          filter,
+        }
+      );
+    let tokenObjects: Array<SavedObjectsFindResult<UninstallTokenSOAttributes>> = [];
+    for await (const result of tokensFinder.find()) {
+      tokenObjects = [...tokenObjects, ...result.saved_objects];
+    }
+    tokensFinder.close();
+
+    const tokensMap = tokenObjects.reduce((acc, { attributes }) => {
+      const policyId = attributes.policy_id;
+      const token = attributes.token || attributes.token_plain;
+      if (!policyId || !token) {
+        return acc;
+      }
+
+      return {
+        ...acc,
+        [policyId]: token,
+      };
+    }, {} as Record<string, string>);
+
+    return tokensMap;
+  }
+
+  /**
+   * gets uninstall tokens for all policies
+   *
+   * @returns Record<policyId, token>
+   */
+  public async getAllTokens(): Promise<Record<string, string>> {
+    const policyIds = await this.getAllPolicyIds();
+    return this.getTokensForPolicyIds(policyIds);
+  }
+
+  /**
+   * get hashed uninstall token for given policy id
+   *
+   * @param policyId agent policy id
+   * @returns hashedToken
+   */
+  public async getHashedTokenForPolicyId(policyId: string): Promise<string> {
+    return (await this.getHashedTokensForPolicyIds([policyId]))[policyId];
+  }
+
+  /**
+   * get hashed uninstall tokens for given policy ids
+   *
+   * @param policyIds agent policy ids
+   * @returns Record<policyId, hashedToken>
+   */
+  public async getHashedTokensForPolicyIds(policyIds: string[]): Promise<Record<string, string>> {
+    const tokensMap = await this.getTokensForPolicyIds(policyIds);
+    return Object.entries(tokensMap).reduce((acc, [policyId, token]) => {
+      if (!policyId || !token) {
+        return acc;
+      }
+      return { ...acc, [policyId]: this.hashToken(token) };
+    }, {});
+  }
+
+  /**
+   * get hashed uninstall token for all policies
+   *
+   * @returns Record<policyId, hashedToken>
+   */
+  public async getAllHashedTokens(): Promise<Record<string, string>> {
+    const policyIds = await this.getAllPolicyIds();
+    return this.getHashedTokensForPolicyIds(policyIds);
+  }
+
+  /**
+   * generate uninstall token for given policy id
+   * will not create a new token if one already exists for a given policy unless force: true is used
+   *
+   * @param policyId agent policy id
+   * @param force generate a new token even if one already exists
+   * @returns hashedToken
+   */
+  public async generateTokenForPolicyId(policyId: string, force: boolean = false): Promise<string> {
+    return (await this.generateTokensForPolicyIds([policyId], force))[policyId];
+  }
+
+  /**
+   * generate uninstall tokens for given policy ids
+   * will not create a new token if one already exists for a given policy unless force: true is used
+   *
+   * @param policyIds agent policy ids
+   * @param force generate a new token even if one already exists
+   * @returns Record<policyId, hashedToken>
+   */
+  public async generateTokensForPolicyIds(
+    policyIds: string[],
```

Comment:
I'm wondering if, for the case we query all the policy Id's, we could filter to get all policy id's without a token. I'm specially wondering when running the `generateTokensForAllPolicies` from fleet setup. That means on every setup we are gonna query all the policies and tokens to check which ones need to be generated. If we query on ES only for policies without a token generated yet, we could avoid a lot of work I guess. Thoughts?

Reply:
we're doing this below:
```
const existingTokens = force ? {} : await this.getTokensForPolicyIds(policyIds);
const missingTokenPolicyIds = force
  ? policyIds
  : policyIds.filter((policyId) => !existingTokens[policyId]);
```

we can't filter it out at policyId query time since the token doesn't live in that document.